compiler/rustc_trait_selection/src/solve/canonicalize.rs

```diff
@@ -202,8 +204,18 @@ impl<'tcx> TypeFolder<TyCtxt<'tcx>> for Canonicalizer<'_, 'tcx> {
     where
         T: TypeFoldable<TyCtxt<'tcx>>,
     {
+        let vars = t.bound_vars().into_iter().map(|var| match var {
+            BoundVariableKind::Ty(_) => BoundVariableKind::Ty(BoundTyKind::Anon),
+            BoundVariableKind::Region(_) => {
+                // FIXME(-Ztrait-solver=next): is it fine to canonicalize away BrEnv
+                BoundVariableKind::Region(BoundRegionKind::BrAnon(None))
+            }
+            BoundVariableKind::Const => BoundVariableKind::Const,
+        });
+        let vars = self.infcx.tcx.mk_bound_variable_kinds_from_iter(vars);
+
         self.binder_index.shift_in(1);
-        let t = t.super_fold_with(self);
+        let t = ty::Binder::bind_with_vars(t.skip_binder().fold_with(self), vars);
```

Comment:
Reviewing from my phone, otherwise I would just test -- this is renumbering the bound var list but not actually changing the references to that var in the binder itself, right?

Also, I wonder if this should just be using https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.anonymize_bound_vars